### PR TITLE
Refactor mwApiPath (to mwActionApiPath), mwRestApiPath, mwModulePath and mwWikiPath parameters + fix regression of test render template

### PR DESCRIPTION
**Pull request by VadimKovalenkoSNF**
Fixes https://github.com/openzim/mwoffliner/issues/1935
Depends on https://github.com/openzim/mwoffliner/pull/1944

List of fixes:

1. Comment out VisualEditor capabilities check and tests. It should be enabled back once https://phabricator.wikimedia.org/T350117 is unblocked.
2. Partially comment out  error checking in checkApiAvailability till the issue with VE is resolved.
3. Fixed regressions around **mwApiPath**, **mwRestApiPath**, **mwModulePath**, **mwWikiPath**. Updated base director methods.
4. Renamed **mwApiPath** to **mwActionApiPath**.
5. ~~Explicitly set `{ mwWikiPath: '/' }` in the tests to prevent default **wiki/** path to assigning.~~ ~~Default value of `mwWikiPath` is the empty string as per discussion in https://github.com/openzim/mwoffliner/issues/1935#issuecomment-1803449460~~ Default wikiPath is 'wiki/' as it was before.

~~:exclamation: Important - this will probably affect receipts in the zimfarm because `--mwWikiPath='/'` is obligated param for the Mediawiki wikis now.~~

~~Check this command as an example: `npm start -- --mwUrl=https://en.wikipedia.org --adminEmail=mail@mail.com --articleList=Canada --mwWikiPath=/`.~~

~~See https://github.com/openzim/mwoffliner/issues/1935#issuecomment-1786756538 for more details.~~
^ Upd: this is no longer valid, mwWikiPath can be default 'wiki/'

6. Refactored regression with test/testAllRenders.ts  - fixed renderer handling across e2e tests, fixed issue with non-removed test zim files, added optional render list that is useful for wikis that do not support all renders from RENDERERS_LIST
7. Optimized src/renderers/renderer.builder.ts to meet Codecov requirements.
8. Fixed Mediawiki setters side effects when they received undefined parameters.

### CODE REVIEW

**Comment by kelson42** [comment]
> * Comment out VisualEditor capabilities check and tests. It should be enabled back once https://phabricator.wikimedia.org/T350117 is unblocked.
> 
> * Partially comment out  error checking in checkApiAvailability till the issue with VE is resolved.

This PR has nothing to do with VE, as such nothing regarding VE should normaly be inside... and for sure not https://phabricator.wikimedia.org/T350117 which has really nothing to do with this ticket.

File: src/util/builders/url/base.director.ts [hunk 1]
```diff
@@ -10,7 +10,13 @@ export default class BaseURLDirector {
     this.baseDomain = baseDomain
   }
 
-  buildURL(path: string) {
+  buildURL(path: string, wikiPath = '') {
```
Comment (kelson42) [inline src/util/builders/url/base.director.ts:13]: command line inputs should be sanitize in `sanitize-argument.ts`

File: src/Downloader.ts [hunk 2]
```diff
@@ -175,13 +176,13 @@ class Downloader {
       this.baseUrl = basicURLDirector.buildDownloaderBaseUrl([
         { condition: await MediaWiki.hasWikimediaMobileApi(), value: MediaWiki.WikimediaMobileApiUrl.href },
         { condition: await MediaWiki.hasWikimediaDesktopApi(), value: MediaWiki.WikimediaDesktopApiUrl.href },
-        { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.visualEditorApiUrl.href },
+        // { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.visualEditorApiUrl.href },
```
Comment (kelson42) [inline src/Downloader.ts:179]: 1 - create a PR to deactivate visualeditor
2 - create a ticket requesting reintroducing it once the phabricator ist fix and rolled out
3 - put links between all of this so things are understandable

Only the VisualEditor autodetection should be deactivated

File: test/e2e/multimediaContent.test.ts [hunk 3]
```diff
@@ -97,12 +70,143 @@ await testAllRenders({ ...parameters, format: ['nopic', 'novid', 'nopdf', 'nodet
...
+                  ].sort(),
+                )
+              }
+            }
+          })
+          afterAll(() => {
+            rimraf.sync(`./${outFiles[0].testId}`)
+          })
+          break
+        case 'VisaulEditor':
```
Comment (kelson42) [inline test/e2e/multimediaContent.test.ts:150]: typo

**Review by kelson42 — CHANGES_REQUESTED** [review]
Few changes to do only.

File: src/mwoffliner.lib.ts [hunk 4]
```diff
@@ -75,8 +75,8 @@ async function execute(argv: any) {
     keepEmptyParagraphs,
     mwUrl,
     mwWikiPath,
-    mwActionApiPath,
     mwApiPath,
```
Comment (kelson42) [inline src/mwoffliner.lib.ts:78]: The `mwApiPath` is ambiguous, now that MW provides many API end-points. Please rename `mwApiPath` to `mwActionApiPath` everywhere (not only on the command line).

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/mwoffliner.lib.ts (hunk 4 above)
Comment (VadimKovalenkoSNF) [reply src/mwoffliner.lib.ts]: Done.

File: test/e2e/multimediaContent.test.ts (hunk 3 above)
Comment (VadimKovalenkoSNF) [reply test/e2e/multimediaContent.test.ts]: Fixed.

File: src/Downloader.ts (hunk 2 above)
Comment (VadimKovalenkoSNF) [reply src/Downloader.ts]: Done, this PR has been merged - https://github.com/openzim/mwoffliner/pull/1941
Ticket about reintroducing VE - https://github.com/openzim/mwoffliner/issues/1940

File: src/util/builders/url/base.director.ts (hunk 1 above)
Comment (VadimKovalenkoSNF) [reply src/util/builders/url/base.director.ts:13]: Done.

**Comment by benoit74** [comment]
I like the fact that you completely removed `--mwApiPath` CLI param. Reusing it for the REST endpoint (as it is currently done in `main` branch) was an extremely breaking change from my PoV.

This CLI changes are somehow a breaking change (we cannot set `--mwApiPath` anymore) but at least we will get a clear error instead of very weird things when the parameter was finally used for something else.

Good job!

**Comment by VadimKovalenkoSNF** [comment]
Thank you, @benoit74 , but I want to emphasize that `mwApiPath` was renamed to `mwActionApiPath`. Full list of param definitions can be found in `src/parameterList.ts`. Moreover, `mwWikiPath` is obligated now (this basically was introduced in this PR) for scraping wikimedia wikis correctly and `mwWikiPath` and `mwActionApiPath` are necessary for 3rd party wikis, see https://github.com/openzim/mwoffliner/issues/1935#issuecomment-1801338689 for more details.

**Review by kelson42 — CHANGES_REQUESTED** [review]
CI does not pass

**Comment by VadimKovalenkoSNF** [comment]
> CI does not pass

Because of the regression described in https://github.com/openzim/mwoffliner/issues/1943

**Review by kelson42 — CHANGES_REQUESTED** [review]

**Comment by VadimKovalenkoSNF** [comment]
@kelson42  It looks like the root cause was not with treating `wikiPath` for the third-party wikis, but rather how `mwActionApi` (ex mwActionApi) is set - it should not contain the forward slash at the beginning, because we build API endpoint on top of baseUrl.href in Mediawiki which already contains forward slash at the end.

Check this JS snippet:
```
const url = 'https://en.wikipedia.org'
console.log(new URL(url).href) // "https://en.wikipedia.org/" 
```
Forward slash at the end of href prohibits building the correct URL for wikis like `https://minecraft.wiki/`: it will be like `https://minecraft.wiki//api.php?action=...` which is wrong (note two forward slashes before `api.php/`). So in such cases, the receipt should be changed. Here is a minified version of article scraping when using changes from this PR:
`npm start -- --adminEmail="contact@kiwix.org" --customZimDescription="Minecraft is a sandbox construction video game." --mwActionApiPath="api.php" --mwModulePath="/load.php" --mwUrl="https://minecraft.wiki/" --mwWikiPath="/" --articleList="Enchanting" --verbose`

By using this command, the action API URL to get wiki site info will be correct:
https://minecraft.wiki/api.php?action=query&meta=siteinfo&format=json&formatversion=2&siprop=general%7Cnamespaces%7Cstatistics%7Cvariables%7Ccategory%7Cwikidesc

So I brought back the default 'wiki/' param for Mediawiki.webUrl. It is using only there and nowhere else.

File: src/MediaWiki.ts [hunk 5]
```diff
@@ -178,23 +205,25 @@ class MediaWiki {
...
+    this.WikimediaMobileApiUrl = this.baseUrlDirector.buildWikimediaMobileApiUrl(this.#restApiPath)
+    this.mobileModulePath = this.baseUrlDirector.buildMobileModuleURL()
     this.wikimediaDesktopUrlDirector = new WikimediaDesktopURLDirector(this.WikimediaDesktopApiUrl.href)
     this.wikimediaMobileUrlDirector = new WikimediaMobileURLDirector(this.WikimediaMobileApiUrl.href)
-    this.VisualEditorURLDirector = new VisualEditorURLDirector(this.visualEditorApiUrl.href)
+  }
+
+  private initApiURLDirector() {
+    // TODO: this.webUrl probably shouldn't accept hardcoded 'wiki/' param, check test/e2e/extra.e2e.test.ts
+    this.webUrl = this.baseUrlDirector.buildURL('wiki/')
```
[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply src/MediaWiki.ts]: Refactored this. Also, 'wiki/' value is default again.

File: test/e2e/forceRender.test.ts [hunk 6]
```diff
@@ -39,7 +40,7 @@ describe('forceRender', () => {
     expect(redisScan.stdout).toEqual('')
   })
 
-  test('Scrape article from bm.wikipedia.org should throw error when using VisualEditor render', async () => {
+  test.skip('Scrape article from bm.wikipedia.org should throw error when using VisualEditor render', async () => {
```
[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply test/e2e/forceRender.test.ts]: Fixed.

File: test/testAllRenders.ts [hunk 7]
```diff
@@ -42,19 +47,25 @@ async function checkZimTools() {
...
 async function getOutFiles(renderName: string, testId: string, parameters: Parameters): Promise<any> {
   await execa('redis-cli flushall', { shell: true })
-  const outFiles = await mwoffliner.execute(parameters)
+  const outFiles = await mwoffliner.execute({ ...parameters, outputDirectory: testId, forceRender: renderName })
 
   return outFiles
 }
 
-export async function testAllRenders(parameters: Parameters, callback) {
+export async function testAllRenders(parameters: Parameters, callback, optionalRenderesList?: Array<string>) {
```
[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply test/testAllRenders.ts]: Refactored.

File: test/unit/styles.test.ts [hunk 8]
```diff
@@ -13,7 +13,7 @@ describe('Styles', () => {
 
   test('Stylesheet downloading', async () => {
     const { articleDetailXId } = RedisStore
-    const { downloader } = await setupScrapeClasses() // en wikipedia
+    const { downloader } = await setupScrapeClasses({ mwWikiPath: '/' }) // en wikipedia
```
[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply test/unit/styles.test.ts]: I fixed this issue.

**Pull request merged**